# **Music Recommendation System**

# **Milestone 2**

Now that we have explored the data, let's apply different algorithms to build recommendation systems.

**Note:** Use the shorter version of the data, i.e., the data after the cutoffs as used in Milestone 1.

In [1]:
# Used to ignore the warning given as output of the code
import warnings
warnings.filterwarnings('ignore')

# Basic libraries of python for numeric and dataframe computations
import numpy as np
import pandas as pd

# Basic library for data visualization
import matplotlib.pyplot as plt

# Slightly advanced library for data visualization
import seaborn as sns

# To compute the cosine similarity between two vectors
from sklearn.metrics.pairwise import cosine_similarity

# A dictionary output that does not raise a key error
from collections import defaultdict

# A performance metrics in sklearn
from sklearn.metrics import mean_squared_error

## **Load the dataset**

In [2]:
# count_data = pd.read_csv('count_data.csv')
# song_data = pd.read_csv('song_data.csv')
# Load the dataset you have saved at the end of milestone 1
df_final = pd.read_csv('df_final.csv')

In [3]:
df_final.head()

user_id  song_id  play_count                            title  \
0     6958      447           1               Daisy And Prudence   
1     6958      512           1  The Ballad of Michael Valentine   
2     6958      549           1        I Stand Corrected (Album)   
3     6958      703           1            They Might Follow You   
4     6958      719           1                       Monkey Man   

                release      artist_name  year  
0          Distillation     Erin McKeown  2000  
1               Sawdust      The Killers  2004  
2       Vampire Weekend  Vampire Weekend  2007  
3           Tiny Vipers      Tiny Vipers  2007  
4  You Know I'm No Good    Amy Winehouse  2007

In [4]:
df_final.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 117876 entries, 0 to 117875
Data columns (total 7 columns):
 #   Column       Non-Null Count   Dtype 
---  ------       --------------   ----- 
 0   user_id      117876 non-null  int64 
 1   song_id      117876 non-null  int64 
 2   play_count   117876 non-null  int64 
 3   title        117876 non-null  object
 4   release      117876 non-null  object
 5   artist_name  117876 non-null  object
 6   year         117876 non-null  int64 
dtypes: int64(4), object(3)
memory usage: 6.3+ MB


### **Popularity-Based Recommendation Systems**

Let's take the count and sum of play counts of the songs and build the popularity recommendation systems based on the sum of play counts.

In [5]:
# Calculating average play_count
average_count = df_final.groupby('song_id').mean()['play_count']      # Hint: Use groupby function on the song_id column

# Calculating the frequency a song is played
play_freq = df_final.groupby('song_id').count()['play_count']        # Hint: Use groupby function on the song_id column

In [6]:
# Making a dataframe with the average_count and play_freq
final_play = pd.DataFrame({'avg_count':average_count, 'play_freq':play_freq})

# Let us see the first five records of the final_play dataset
final_play.head()

avg_count  play_freq
song_id                      
21        1.622642        265
22        1.492424        132
52        1.729216        421
62        1.728070        114
93        1.452174        115

Now, let's create a function to find the top n songs for a recommendation based on the average play count of song. We can also add a threshold for a minimum number of playcounts for a song to be considered for recommendation.

In [7]:
# Build the function to find top n songs
def top_n_songs(data, n, min_interaction = 100):
    
    # Finding movies with interactions greater than the minimum number of interactions
    recommendations = data[data['play_freq'] > min_interaction]
    
    # Sorting values with respect to the average rating
    recommendations = recommendations.sort_values(by = 'avg_count', ascending = False)
    
    return recommendations.index[:n]

In [8]:
# Recommend top 10 songs using the function defined above
list(top_n_songs(final_play, 10, 100))

[7224, 6450, 9942, 5531, 5653, 8483, 2220, 657, 614, 352]

### **User User Similarity-Based Collaborative Filtering**

To build the user-user-similarity-based and subsequent models we will use the "surprise" library.

In [9]:
# Install the surprise package using pip. Uncomment and run the below code to do the same
!pip install surprise 

In [10]:
# Import necessary libraries

# To compute the accuracy of models
from surprise import accuracy

# This class is used to parse a file containing play_counts, data should be in structure - user; item; play_count
from surprise.reader import Reader

# Class for loading datasets
from surprise.dataset import Dataset

# For tuning model hyperparameters
from surprise.model_selection import GridSearchCV

# For splitting the data in train and test dataset
from surprise.model_selection import train_test_split

# For implementing similarity-based recommendation system
from surprise.prediction_algorithms.knns import KNNBasic

# For implementing matrix factorization based recommendation system
from surprise.prediction_algorithms.matrix_factorization import SVD

# For implementing KFold cross-validation
from surprise.model_selection import KFold

# For implementing clustering-based recommendation system
from surprise import CoClustering

### Some useful functions

Below is the function to calculate precision@k and recall@k, RMSE and F1_Score@k to evaluate the model performance.

**Think About It:** Which metric should be used for this problem to compare different models?

In [11]:
# The function to calulate the RMSE, precision@k, recall@k, and F_1 score
def precision_recall_at_k(model, k = 30, threshold = 1.5):
    """Return precision and recall at k metrics for each user"""

    # First map the predictions to each user.
    user_est_true = defaultdict(list)
    
    # Making predictions on the test data
    predictions=model.test(testset)
    
    for uid, _, true_r, est, _ in predictions:
        user_est_true[uid].append((est, true_r))

    precisions = dict()
    recalls = dict()
    for uid, user_ratings in user_est_true.items():

        # Sort user ratings by estimated value
        user_ratings.sort(key = lambda x : x[0], reverse = True)

        # Number of relevant items
        n_rel = sum((true_r >= threshold) for (_, true_r) in user_ratings)

        # Number of recommended items in top k
        n_rec_k = sum((est >= threshold) for (est, _) in user_ratings[ : k])

        # Number of relevant and recommended items in top k
        n_rel_and_rec_k = sum(((true_r >= threshold) and (est >= threshold))
                              for (est, true_r) in user_ratings[ : k])

        # Precision@K: Proportion of recommended items that are relevant
        # When n_rec_k is 0, Precision is undefined. We here set Precision to 0 when n_rec_k is 0

        precisions[uid] = n_rel_and_rec_k / n_rec_k if n_rec_k != 0 else 0

        # Recall@K: Proportion of relevant items that are recommended
        # When n_rel is 0, Recall is undefined. We here set Recall to 0 when n_rel is 0

        recalls[uid] = n_rel_and_rec_k / n_rel if n_rel != 0 else 0
    
    # Mean of all the predicted precisions are calculated
    precision = round((sum(prec for prec in precisions.values()) / len(precisions)), 3)

    # Mean of all the predicted recalls are calculated
    recall = round((sum(rec for rec in recalls.values()) / len(recalls)), 3)
    
    accuracy.rmse(predictions)

    # Command to print the overall precision
    print('Precision: ', precision)

    # Command to print the overall recall
    print('Recall: ', recall)
    
    # Formula to compute the F-1 score
    print('F_1 score: ', round((2 * precision * recall) / (precision + recall), 3))

**Think About It:** In the function precision_recall_at_k above the threshold value used is 1.5. How precision and recall are affected by changing the threshold? What is the intuition behind using the threshold value of 1.5? 

In [12]:
# Instantiating Reader scale with expected rating scale 
reader = Reader(rating_scale=(0, 5)) #use rating scale (0, 5)

# Loading the dataset
data = Dataset.load_from_df(df_final[["user_id", "song_id", "play_count"]], reader) # Take only "user_id","song_id", and "play_count"

# Splitting the data into train and test dataset
trainset, testset = train_test_split(data, test_size=0.4, random_state = 42) # Take test_size = 0.4

**Think About It:** How changing the test size would change the results and outputs?

In [13]:
# Build the default user-user-similarity model
sim_options = {'name': 'cosine',
               'user_based':True}

# KNN algorithm is used to find desired similar items
sim_user_user = KNNBasic(sim_options = sim_options, verbose = False, random_state = 1) # Use random_state = 1 

# Train the algorithm on the trainset, and predict play_count for the testset
sim_user_user.fit(trainset)

# Let us compute precision@k, recall@k, and f_1 score with k = 30
precision_recall_at_k(sim_user_user) # Use sim_user_user model

RMSE: 1.0878
Precision:  0.396
Recall:  0.692
F_1 score:  0.504


**Observations and Insights:**
- We have calculated **RMSE** to check **how far the overall predicted ratings** are from the **actual ratings**.
- Intuition of Recall - We are getting a **recall of almost 0.692**, which means out of **all the relevant songs, 69.2% are recommended**. 
- Intuition of Precision - We are getting a **precision of almost 0.396**, which means **out of all the recommended songs 39.6% are relevant**.
- Here **F_1 score** of the **baseline model is almost 0.504**. It indicates that **mostly recommended songs were relevant and relevant songs were recommended** to the user. We will try to improve this later by using **GridSearchCV by tuning different hyperparameters** of this algorithm.

In [14]:
# Predicting play_count for a sample user with a listened song
sim_user_user.predict(6958, 1671, r_ui = 5, verbose = True) # Use user id 6958 and song_id 1671

user: 6958       item: 1671       r_ui = 5.00   est = 1.80   {'actual_k': 40, 'was_impossible': False}


Prediction(uid=6958, iid=1671, r_ui=5, est=1.8009387435128914, details={'actual_k': 40, 'was_impossible': False})

In [15]:
# Predicting play_count for a sample user with a song not-listened by the user
sim_user_user.predict(6958, 3232, r_ui = 5, verbose = True) # Use user_id 6958 and song_id 3232

user: 6958       item: 3232       r_ui = 5.00   est = 1.64   {'actual_k': 40, 'was_impossible': False}


Prediction(uid=6958, iid=3232, r_ui=5, est=1.6386860897998294, details={'actual_k': 40, 'was_impossible': False})

**Observations and Insights:**
- The above output shows that **the actual rating for this user-item pair is 5 and the predicted rating is 1.80** by the **user-user-similarity-based baseline model**. This implies that the model is under-estimating the ratings.
- The above output shows that **the actual rating for this user-item pair is 5 and the predicted rating is 1.64** by the **user-user-similarity-based baseline model**. This implies that the model is under-estimating the ratings.

Now, let's try to tune the model and see if we can improve the model performance.

In [16]:
# Setting up parameter grid to tune the hyperparameters
param_grid = {'k': [10, 20, 30], 'min_k': [3, 6, 9],
              'sim_options': {'name': ["cosine", 'pearson', "pearson_baseline"],
                              'user_based': [True], "min_support": [2, 4]}
              }

# Performing 3-fold cross-validation to tune the hyperparameters
gs = GridSearchCV(KNNBasic, param_grid, measures = ['rmse'], cv = 3, n_jobs = -1)

# Fitting the data
gs.fit(data) # Use entire data for GridSearch

# Best RMSE score
print(gs.best_score['rmse'])

# Combination of parameters that gave the best RMSE score
print(gs.best_params['rmse'])

1.0453958972841557
{'k': 30, 'min_k': 9, 'sim_options': {'name': 'pearson_baseline', 'user_based': True, 'min_support': 2}}


In [17]:
# Train the best model found in above gridsearch
sim_options = {'name': 'pearson_baseline',
               'user_based': True,
               'min_support': 2}

# Creating an instance of KNNBasic with optimal hyperparameter values
sim_user_user_optimized = KNNBasic(sim_options = sim_options, k = 30, min_k = 9, random_state = 1, verbose = False)

# Training the algorithm on the trainset
sim_user_user_optimized.fit(trainset)

# Let us compute precision@k, recall@k, and F_1 score with k = 10
precision_recall_at_k(sim_user_user_optimized)

RMSE: 1.0521
Precision:  0.413
Recall:  0.721
F_1 score:  0.525


**Observations and Insights:**
- We can see from above that after tuning hyperparameters, **F_1 score of the tuned model has reduced a bit as compared to the baseline model.**

In [18]:
# Predict the play count for a user who has listened to the song. Take user_id 6958, song_id 1671 and r_ui = 2
sim_user_user_optimized.predict(6958, 1671, r_ui = 2, verbose = True)

user: 6958       item: 1671       r_ui = 2.00   est = 1.96   {'actual_k': 24, 'was_impossible': False}


Prediction(uid=6958, iid=1671, r_ui=2, est=1.962926073914969, details={'actual_k': 24, 'was_impossible': False})

In [19]:
songs_interacted_by_6958 = df_final[df_final.user_id == 6958].song_id.unique()
songs_interacted_by_6958

array([ 447,  512,  549,  703,  719,  892, 1050, 1480, 1671, 1752, 1756,
       1787, 1818, 2107, 2289, 2304, 2425, 2501, 2701, 2898, 2994, 3074,
       3491, 3551, 3718, 3801, 3907, 5193, 5340, 5441, 5566, 5894, 6305,
       7738, 8029, 8037, 8425, 9065, 9351], dtype=int64)

In [20]:
all_songs = df_final.song_id.unique()

In [21]:
def non_match_elements(list_a, list_b):
    non_match = []
    for i in list_a:
        if i not in list_b:
            non_match.append(i)
    return non_match

In [22]:
# Songs that are not listened to by the user (with user_id 6958)
non_match = non_match_elements(all_songs, songs_interacted_by_6958)
non_match[:5]

[97, 122, 139, 693, 765]

In [94]:
df_final[df_final.user_id == 6958].song_id.unique()

array([ 447,  512,  549,  703,  719,  892, 1050, 1480, 1671, 1752, 1756,
       1787, 1818, 2107, 2289, 2304, 2425, 2501, 2701, 2898, 2994, 3074,
       3491, 3551, 3718, 3801, 3907, 5193, 5340, 5441, 5566, 5894, 6305,
       7738, 8029, 8037, 8425, 9065, 9351], dtype=int64)

In [23]:
# Predict the play count for a song that is not listened to by the user (with user_id 6958)
sim_user_user_optimized.predict(6958, 97, r_ui = 2, verbose = True)

user: 6958       item: 97         r_ui = 2.00   est = 1.55   {'actual_k': 30, 'was_impossible': False}


Prediction(uid=6958, iid=97, r_ui=2, est=1.5527290962841591, details={'actual_k': 30, 'was_impossible': False})

**Observations and Insights:**
- The above output shows that **the actual rating for this user-item pair is 2 and the predicted rating is 1.70** by the **user-user-similarity-based optimized model**. This implies that the model is under-estimating the ratings.
- The above output shows that **the actual rating for this user-item pair is 2 and the predicted rating is 1.55** by the **user-user-similarity-based optimized model**. This implies that the model is under-estimating the ratings.

**Think About It:** Along with making predictions on listened and unknown songs can we get 5 nearest neighbors (most similar) to a certain song?

In [24]:
# Use inner id 0


Below we will be implementing a function where the input parameters are:

- data: A **song** dataset
- user_id: A user-id **against which we want the recommendations**
- top_n: The **number of songs we want to recommend**
- algo: The algorithm we want to use **for predicting the play_count**
- The output of the function is a **set of top_n items** recommended for the given user_id based on the given algorithm

In [25]:
def get_recommendations(data, user_id, top_n, algo):
    
    # Creating an empty list to store the recommended product ids
    recommendations = []
    
    # Creating an user item interactions matrix 
    user_item_interactions_matrix = data.pivot_table(index = 'user_id', columns = 'song_id', values = 'play_count')
    
    # Extracting those business ids which the user_id has not visited yet
    non_interacted_products = user_item_interactions_matrix.loc[user_id][user_item_interactions_matrix.loc[user_id].isnull()].index.tolist()
    
    # Looping through each of the business ids which user_id has not interacted yet
    for item_id in non_interacted_products:
        
        # Predicting the ratings for those non visited restaurant ids by this user
        est = algo.predict(user_id, item_id).est
        
        # Appending the predicted ratings
        recommendations.append((item_id, est))

    # Sorting the predicted ratings in descending order
    recommendations.sort(key = lambda x : x[1], reverse = True)

    return recommendations[:top_n] # Returing top n highest predicted rating products for this user

In [26]:
# Make top 5 recommendations for user_id 6958 with a similarity-based recommendation engine
recommendations = get_recommendations(df_final, 6958, 5, sim_user_user_optimized)

In [27]:
# Building the dataframe for above recommendations with columns "song_id" and "predicted_ratings"
pd.DataFrame(recommendations, columns = ['song_id', 'predicted_ratings'])

song_id  predicted_ratings
0     5531           2.553335
1      317           2.518269
2     4954           2.406776
3     8635           2.396606
4     5943           2.390723

**Observations and Insights:**
While comparing the ratings of two songs, it is not only the **ratings** that describe the **likelihood of the user to that song**. Along with the rating the **number of users who have listened to that song** also becomes important to consider. Due to this, we have calculated the **"predicted_ratings"** for each song. Commonly higher the **"play_count" of a song more it is liked by users**. To interpret the above concept, a **song rated 4 by 3 people is less liked in comparison to a song rated 3 by 50 people**. It has been **empirically found that the likelihood of the song is directly proportional to the inverse of the square root of the play_count of the song**.

### Correcting the play_counts and Ranking the above songs

In [28]:
def ranking_songs(recommendations, final_rating):
  # Sort the songs based on play counts
  ranked_songs = final_rating.loc[[items[0] for items in recommendations]].sort_values('play_freq', ascending = False)[['play_freq']].reset_index()

  # Merge with the recommended songs to get predicted play_count
  ranked_songs = ranked_songs.merge(pd.DataFrame(recommendations, columns = ['song_id', 'predicted_ratings']), on = 'song_id', how = 'inner')

  # Rank the songs based on corrected play_counts
  ranked_songs['corrected_ratings'] = ranked_songs['predicted_ratings'] - 1 / np.sqrt(ranked_songs['play_freq'])

  # Sort the songs based on corrected play_counts
  ranked_songs = ranked_songs.sort_values('corrected_ratings', ascending = False)
  
  return ranked_songs

**Think About It:** In the above function to correct the predicted play_count a quantity 1/np.sqrt(n) is subtracted. What is the intuition behind it? Is it also possible to add this quantity instead of subtracting?

In [29]:
# Applying the ranking_songs function on the final_play data
ranking_songs(recommendations, final_play)

song_id  play_freq  predicted_ratings  corrected_ratings
0     5531        618           2.553335           2.513109
2      317        411           2.518269           2.468943
1     5943        423           2.390723           2.342101
3     4954        183           2.406776           2.332854
4     8635        155           2.396606           2.316284

**Observations and Insights:**
- The above output shows that **the corrected ratings are higher than the predicted ratings** by the **user-user-similarity-based optimized model**. This implies that the model is over-estimating the ratings.

### Item Item Similarity-based collaborative filtering recommendation systems 

In [30]:
# Apply the item-item similarity collaborative filtering model with random_state = 1 and evaluate the model performance
sim_options = {'name': 'cosine',
               'user_based': False}

# KNN algorithm is used to find desired similar items
sim_item_item = KNNBasic(sim_options = sim_options, random_state = 1, verbose = False)

# Train the algorithm on the trainset, and predict ratings for the testset
sim_item_item.fit(trainset)

# Let us compute precision@k, recall@k, and f_1 score with k = 10
precision_recall_at_k(sim_item_item)

RMSE: 1.0394
Precision:  0.307
Recall:  0.562
F_1 score:  0.397


**Observations and Insights:**
- We have calculated **RMSE** to check **how far the overall predicted ratings** are from the **actual ratings**.
- Intuition of Recall - We are getting a **recall of almost 0.562**, which means out of **all the relevant songs, 56.2% are recommended**. 
- Intuition of Precision - We are getting a **precision of almost 0.307**, which means **out of all the recommended songs 30.7% are relevant**.
- Here **F_1 score** of the **baseline model is almost 0.397**. It indicates that **mostly recommended songs were relevant and relevant songs were recommended** to the user. We will try to improve this later by using **GridSearchCV by tuning different hyperparameters** of this algorithm.

In [31]:
# Predicting play count for a sample user_id 6958 and song (with song_id 1671) heard by the user
sim_item_item.predict(6958, 1671, r_ui = 5, verbose = True)

user: 6958       item: 1671       r_ui = 5.00   est = 1.36   {'actual_k': 20, 'was_impossible': False}


Prediction(uid=6958, iid=1671, r_ui=5, est=1.3614157231762556, details={'actual_k': 20, 'was_impossible': False})

In [32]:
song1671_interacted_by_users = df_final[df_final.song_id == 1671].user_id.unique()
song1671_interacted_by_users

array([ 6958, 45386, 22749, 51415, 74334, 15700, 18349, 33280, 41799,
       54866, 23516, 21117, 63702, 27240, 14538, 69334, 44790, 45349,
       70987, 53145, 11453, 49747,  6141, 20797, 30305, 55566, 57374,
        1035, 52134, 23705, 60129, 58757, 11010, 18409, 28007, 59672,
        7670, 20809, 25351, 15644, 21045, 68061, 35160, 55323, 31451,
        4848, 23790, 41878, 69900, 53691, 43177, 52662, 72448, 60644,
       60518, 40802, 34410, 17434, 56574, 62989, 15634, 71179, 50849,
       63669, 71481, 75934, 38654, 61513, 29038, 59307, 42216, 66345,
       25298,  9975, 52400, 15984, 25051,  2396, 71060,  1604, 33403,
       60008, 39281, 23777, 61358, 48538, 35101, 44726, 23099, 60923,
       70967, 23148, 34385, 69297, 69722, 17793,  1602, 37727, 61794,
       48175, 29786, 72397, 60215, 45152, 45401, 44865, 10267, 43539,
       16915, 72912, 67874, 29527, 64985, 48845, 17932, 48976, 29040,
       40358, 16076, 62379, 11799, 10218, 17050,  9193, 19263, 52192,
       23856, 45335,

In [33]:
all_users = df_final.user_id.unique()
all_users[:5]

array([ 6958, 27018, 55837, 56576, 49549], dtype=int64)

In [34]:
def non_match_elements(list_a, list_b):
    non_match = []
    for i in list_a:
        if i not in list_b:
            non_match.append(i)
    return non_match

In [82]:
# Users who have not listened to the song (with song_id 1671)
non_match = non_match_elements(all_users, song1671_interacted_by_users)
non_match[:5]

[27018, 55837, 56576, 49549, 34234]

In [84]:
# All the songs listended to by user 27018 (note: song 1671 was listined to user 27018)
df_final[df_final.user_id == 27018].song_id.unique()

array([  97,  122,  139,  693,  719,  765,  786,  861, 1118, 1406, 1461,
       1519, 1682, 1696, 1907, 1983, 2187, 2234, 2389, 2557, 3232, 3356,
       3518, 3548, 3655, 4152, 4298, 4377, 4510, 4522, 4545, 4653, 4724,
       4954, 5023, 5158, 5223, 5248, 5305, 5345, 5417, 5457, 5877, 6079,
       6121, 6148, 6191, 6232, 6293, 6350, 6500, 6572, 6618, 6636, 6709,
       6770, 6959, 6964, 7210, 7331, 7399, 7402, 7469, 7721, 7878, 7977,
       8019, 8086, 8247, 8265, 8299, 8434, 8612, 8973, 9055, 9099, 9292,
       9436, 9447, 9476, 9702, 9771, 9989], dtype=int64)

In [85]:
# Predict the play count for a user that has not listened to the song (with song_id 1671)
sim_item_item.predict(27018, 1671, r_ui = 5, verbose = True)

user: 27018      item: 1671       r_ui = 5.00   est = 2.55   {'actual_k': 40, 'was_impossible': False}


Prediction(uid=27018, iid=1671, r_ui=5, est=2.5507234733908, details={'actual_k': 40, 'was_impossible': False})

**Observations and Insights:**
- The above output shows that **the actual rating for this user-item pair is 5 and the predicted rating is 2.55** by the **item-item-similarity-based baseline model**. This implies that the model is under-estimating the ratings.

In [38]:
# Apply grid search for enhancing model performance

# Setting up parameter grid to tune the hyperparameters
param_grid = {'k': [10, 20, 30], 'min_k': [3, 6, 9],
              'sim_options': {'name': ["cosine", 'pearson', "pearson_baseline"],
                              'user_based': [False], "min_support": [2, 4]}
              }

# Performing 3-fold cross-validation to tune the hyperparameters
gs = GridSearchCV(KNNBasic, param_grid, measures = ['rmse'], cv = 3, n_jobs = -1)

# Fitting the data
gs.fit(data)

# Find the best RMSE score
print(gs.best_score['rmse'])

# Extract the combination of parameters that gave the best RMSE score
print(gs.best_params['rmse'])

1.0242402042113319
{'k': 30, 'min_k': 6, 'sim_options': {'name': 'pearson_baseline', 'user_based': False, 'min_support': 2}}


**Think About It:** How do the parameters affect the performance of the model? Can we improve the performance of the model further? Check the list of hyperparameters [here](https://surprise.readthedocs.io/en/stable/knn_inspired.html).

In [39]:
# Apply the best modle found in the grid search
# Using the optimal similarity measure for item-item based collaborative filtering
sim_options = {'name': 'pearson_baseline',
               'user_based': False,
               'min_support': 2}

# Creating an instance of KNNBasic with optimal hyperparameter values
sim_item_item_optimized = KNNBasic(sim_options = sim_options, k = 30, min_k = 6, random_state = 1, verbose = False)

# Training the algorithm on the trainset
sim_item_item_optimized.fit(trainset)

# Let us compute precision@k and recall@k also with k = 10
precision_recall_at_k(sim_item_item_optimized)

RMSE: 1.0328
Precision:  0.408
Recall:  0.665
F_1 score:  0.506


In [93]:
print("sim_item_item_optimized: ")
precision_recall_at_k(sim_item_item_optimized)

sim_item_item_optimized: 
RMSE: 1.0328
Precision:  0.408
Recall:  0.665
F_1 score:  0.506


**Observations and Insights:**
- We can see from above that after tuning hyperparameters, **F_1 score of the tuned model is much better than the baseline model**. Also, there is a considerable fall in the RMSE value after tuning the hyperparameters. Hence the tuned model is doing better than the earlier one.

In [40]:
# Predict the play_count by a user(user_id 6958) for the song (song_id 1671)
sim_item_item_optimized.predict(6958, 1671, r_ui = 2, verbose = True)

user: 6958       item: 1671       r_ui = 2.00   est = 1.96   {'actual_k': 10, 'was_impossible': False}


Prediction(uid=6958, iid=1671, r_ui=2, est=1.9634957386781853, details={'actual_k': 10, 'was_impossible': False})

In [41]:
# Predicting play count for a sample user_id 6958 with song_id 3232 which is not heard by the user
sim_item_item_optimized.predict(6958, 3232, r_ui = 2, verbose = True)

user: 6958       item: 3232       r_ui = 2.00   est = 1.28   {'actual_k': 10, 'was_impossible': False}


Prediction(uid=6958, iid=3232, r_ui=2, est=1.2759946618244609, details={'actual_k': 10, 'was_impossible': False})

**Observations and Insights:**
- The above output shows that **the actual rating for this user-item pair is 2 and the predicted rating is 1.96** by the **item-item-similarity-based optimized model**. This implies that the model is under-estimating the ratings.
- The above output shows that **the actual rating for this user-item pair is 2 and the predicted rating is 1.28** by the **item-item-similarity-based optimized model**. This implies that the model is under-estimating the ratings.

In [42]:
# Find five most similar items to the item with inner id 0
sim_item_item_optimized.get_neighbors(0, k = 5)

[124, 523, 173, 205, 65]

In [43]:
# Making top 5 recommendations for user_id 6958 with item_item_similarity-based recommendation engine
recommendations = get_recommendations(df_final, 6958, 5, sim_item_item_optimized)

In [44]:
# Building the dataframe for above recommendations with columns "song_id" and "predicted_play_count"
pd.DataFrame(recommendations, columns = ['song_id', 'predicted_ratings'])

song_id  predicted_ratings
0     2342           2.653903
1     5101           2.386577
2      139           2.313727
3     7519           2.270864
4     8099           2.212702

In [45]:
# Applying the ranking_songs function
ranking_songs(recommendations, final_play)

song_id  play_freq  predicted_ratings  corrected_ratings
4     2342        111           2.653903           2.558987
2     5101        130           2.386577           2.298871
3      139        119           2.313727           2.222057
1     7519        168           2.270864           2.193712
0     8099        275           2.212702           2.152399

**Observations and Insights:**
- The above output shows that **the predicted ratings are higher than the corrected ratings** by the **item-item-similarity-based optimized model**. This implies that the model is over-estimating the ratings.

### Model Based Collaborative Filtering - Matrix Factorization

Model-based Collaborative Filtering is a **personalized recommendation system**, the recommendations are based on the past behavior of the user and it is not dependent on any additional information. We use **latent features** to find recommendations for each user.

In [46]:
# Build baseline model using svd
svd = SVD(random_state = 1)

# Training the algorithm on the trainset
svd.fit(trainset)

# Let us compute precision@k and recall@k with k = 10
precision_recall_at_k(svd)

RMSE: 1.0252
Precision:  0.41
Recall:  0.633
F_1 score:  0.498


In [47]:
# Making prediction for user (with user_id 6958) to song (with song_id 1671), take r_ui = 2
svd.predict(6958, 1671, r_ui = 2, verbose = True)

user: 6958       item: 1671       r_ui = 2.00   est = 1.27   {'was_impossible': False}


Prediction(uid=6958, iid=1671, r_ui=2, est=1.267473397214638, details={'was_impossible': False})

In [48]:
# Making a prediction for the user who has not listened to the song (song_id 3232)
svd.predict(6958, 3232, r_ui = 2, verbose = True)

user: 6958       item: 3232       r_ui = 2.00   est = 1.56   {'was_impossible': False}


Prediction(uid=6958, iid=3232, r_ui=2, est=1.5561675084403663, details={'was_impossible': False})

#### Improving matrix factorization based recommendation system by tuning its hyperparameters

In [49]:
# Set the parameter space to tune
param_grid = {'n_epochs': [10, 20, 30], 'lr_all': [0.001, 0.005, 0.01],
              'reg_all': [0.2, 0.4, 0.6]}

# Performe 3-fold grid-search cross-validation
gs = GridSearchCV(SVD, param_grid, measures = ['rmse'], cv = 3, n_jobs = -1)

# Fitting data
gs.fit(data)

# Best RMSE score
print(gs.best_score['rmse'])

# Combination of parameters that gave the best RMSE score
print(gs.best_params['rmse'])

1.0134830808863564
{'n_epochs': 30, 'lr_all': 0.01, 'reg_all': 0.2}


**Think About It**: How do the parameters affect the performance of the model? Can we improve the performance of the model further? Check the available hyperparameters [here](https://surprise.readthedocs.io/en/stable/matrix_factorization.html).

In [50]:
# Building the optimized SVD model using optimal hyperparameters
svd_optimized = SVD(n_epochs = 30, lr_all = 0.01, reg_all = 0.2, random_state = 1)

# Training the algorithm on the trainset
svd_optimized = svd_optimized.fit(trainset)

# Let us compute precision@k and recall@k also with k = 10
precision_recall_at_k(svd_optimized)

RMSE: 1.0141
Precision:  0.415
Recall:  0.635
F_1 score:  0.502


**Observations and Insights:**
- We can see from above that after tuning hyperparameters, **F_1 score of the tuned model is a bit better than the baseline model**. Also, there is a small fall in the RMSE value after tuning the hyperparameters. Hence the tuned model is doing better than the earlier one.

In [51]:
# Using svd_optimized model to recommend for userId 6958 and song_id 1671
svd_optimized.predict(6958, 1671, r_ui = 2, verbose = True)

user: 6958       item: 1671       r_ui = 2.00   est = 1.34   {'was_impossible': False}


Prediction(uid=6958, iid=1671, r_ui=2, est=1.3432395286125096, details={'was_impossible': False})

In [52]:
# Using svd_optimized model to recommend for userId 6958 and song_id 3232 with unknown baseline rating
svd_optimized.predict(6958, 3232, r_ui = 2, verbose = True)

user: 6958       item: 3232       r_ui = 2.00   est = 1.44   {'was_impossible': False}


Prediction(uid=6958, iid=3232, r_ui=2, est=1.442548446117648, details={'was_impossible': False})

**Observations and Insights:**
- The above output shows that **the actual rating for this user-item pair is 2 and the predicted rating is 1.34** by the **user-item-matrix factorization based optimized model**. This implies that the model is under-estimating the ratings.
- The above output shows that **the actual rating for this user-item pair is 2 and the predicted rating is 1.44** by the **user-item-matrix factorization based optimized model**. This implies that the model is under-estimating the ratings.

In [53]:
# Getting top 5 recommendations for user_id 6958 using "svd_optimized" algorithm
svd_recommendations = get_recommendations(df_final, 6958, 5, svd_optimized)

In [54]:
# Ranking songs based on above recommendations
ranking_songs(svd_recommendations, final_play)

song_id  play_freq  predicted_ratings  corrected_ratings
2     7224        107           2.601899           2.505225
1     5653        108           2.108728           2.012502
4     8324         96           2.014091           1.912029
0     9942        150           1.940115           1.858465
3     6450        102           1.952493           1.853478

**Observations and Insights:**
- The above output shows that **the predicted ratings are higher than the corrected ratings** by the **user-item-matrix factorization based optimized model**. This implies that the model is over-estimating the ratings.

### Cluster Based Recommendation System

In **clustering-based recommendation systems**, we explore the **similarities and differences** in people's tastes in songs based on how they rate different songs. We cluster similar users together and recommend songs to a user based on play_counts from other users in the same cluster.

In [55]:
# Make baseline clustering model
clust_baseline = CoClustering(random_state = 1)

# Training the algorithm on the trainset
clust_baseline.fit(trainset)

# Let us compute precision@k and recall@k with k = 10
precision_recall_at_k(clust_baseline)

RMSE: 1.0487
Precision:  0.397
Recall:  0.582
F_1 score:  0.472


In [56]:
# Making prediction for user_id 6958 and song_id 1671
clust_baseline.predict(6958, 1671, r_ui = 2, verbose = True)

user: 6958       item: 1671       r_ui = 2.00   est = 1.29   {'was_impossible': False}


Prediction(uid=6958, iid=1671, r_ui=2, est=1.2941824757363074, details={'was_impossible': False})

In [57]:
# Making prediction for user (userid 6958) for a song(song_id 3232) not heard by the user
clust_baseline.predict(6958, 3232, r_ui=2, verbose = True)

user: 6958       item: 3232       r_ui = 2.00   est = 1.48   {'was_impossible': False}


Prediction(uid=6958, iid=3232, r_ui=2, est=1.4785259100797417, details={'was_impossible': False})

#### Improving clustering-based recommendation system by tuning its hyper-parameters

In [58]:
# Set the parameter space to tune
param_grid = {'n_cltr_u': [5, 6, 7, 8], 'n_cltr_i': [5, 6, 7, 8], 'n_epochs': [10, 20, 30]}

# Performing 3-fold grid search cross-validation
gs = GridSearchCV(CoClustering, param_grid, measures = ['rmse'], cv = 3, n_jobs = -1)

# Fitting data
gs.fit(data)

# Best RMSE score
print(gs.best_score['rmse'])

# Combination of parameters that gave the best RMSE score
print(gs.best_params['rmse'])

1.0614632796903807
{'n_cltr_u': 5, 'n_cltr_i': 5, 'n_epochs': 10}


**Think About It**: How do the parameters affect the performance of the model? Can we improve the performance of the model further? Check the available hyperparameters [here](https://surprise.readthedocs.io/en/stable/co_clustering.html).

In [59]:
# Train the tuned Coclustering algorithm
clust_tuned = CoClustering(n_cltr_u = 5,n_cltr_i = 5, n_epochs = 10, random_state = 1)

# Training the algorithm on the trainset
clust_tuned.fit(trainset)

# Let us compute precision@k and recall@k with k = 10
precision_recall_at_k(clust_tuned)

RMSE: 1.0654
Precision:  0.394
Recall:  0.566
F_1 score:  0.465


**Observations and Insights:**
- We can see from above that after tuning hyperparameters, **F_1 score of the tuned model is a bit worse than the baseline model**. Also, there is a small rise in the RMSE value after tuning the hyperparameters. Hence the tuned model is doing worse than the earlier one.

In [60]:
# Using co_clustering_optimized model to recommend for userId 6958 and song_id 1671
clust_tuned.predict(6958, 1671, verbose = True)

user: 6958       item: 1671       r_ui = None   est = 1.91   {'was_impossible': False}


Prediction(uid=6958, iid=1671, r_ui=None, est=1.9108882530486497, details={'was_impossible': False})

In [61]:
# Use Co_clustering based optimized model to recommend for userId 6958 and song_id 3232 with unknown baseline rating
clust_tuned.predict(6958, 3232, verbose = True)

user: 6958       item: 3232       r_ui = None   est = 1.24   {'was_impossible': False}


Prediction(uid=6958, iid=3232, r_ui=None, est=1.2366916027865822, details={'was_impossible': False})

**Observations and Insights:_________**

#### Implementing the recommendation algorithm based on optimized CoClustering model

In [62]:
def get_recommendations(data, user_id, top_n, algo):
    
    # Creating an empty list to store the recommended restaurants ids
    recommendations = []
    
    # Creating an user item interactions matrix 
    user_item_interactions_matrix = data.pivot_table(index = 'user_id', columns = 'song_id', values = 'play_count')
    
    # Extracting those restaurants ids which the user_id has not interacted yet
    non_interacted_restaurants = user_item_interactions_matrix.loc[user_id][user_item_interactions_matrix.loc[user_id].isnull()].index.tolist()
    
    # Looping through each of the restaurants ids which user_id has not interacted yet
    for item_id in non_interacted_restaurants:
        
        # Predicting the ratings for those non interacted restaurants ids by this user
        est = algo.predict(user_id, item_id).est
        
        # Appending the predicted ratings
        recommendations.append((item_id, est))

    # Sorting the predicted ratings in descending order
    recommendations.sort(key = lambda x: x[1], reverse = True)

    # Returing top n highest predicted rating restaurants for this user
    return recommendations[:top_n]

In [63]:
# Getting top 5 recommendations for user_id 6958 using "Co-clustering based optimized" algorithm
clustering_recommendations = get_recommendations(df_final, 6958, 5, clust_tuned)

### Correcting the play_count and Ranking the above songs

In [64]:
def ranking_songs(recommendations, final_rating):
  
  # Sort the restaurants based on ratings count
  ranked_songs = final_rating.loc[[items[0] for items in recommendations]].sort_values('play_freq', ascending = False)[['play_freq']].reset_index()

  # Merge with the recommended restaurants to get predicted ratings
  ranked_songs = ranked_songs.merge(pd.DataFrame(recommendations, columns = ['song_id', 'predicted_ratings']), on = 'song_id', how = 'inner')

  # Rank the restaurants based on corrected ratings
  ranked_songs['corrected_ratings'] = ranked_songs['predicted_ratings'] - 1 / np.sqrt(ranked_songs['play_freq'])

  # Sort the restaurants based on corrected ratings
  ranked_songs = ranked_songs.sort_values('corrected_ratings', ascending = False)
  
  return ranked_songs

In [65]:
# Ranking songs based on the above recommendations
ranking_songs(clustering_recommendations, final_play)

song_id  play_freq  predicted_ratings  corrected_ratings
4     7224        107           3.711503           3.614829
3     5653        108           2.903883           2.807658
0     6860        169           2.691043           2.614120
1      657        151           2.606354           2.524975
2     8483        123           2.582807           2.492640

**Observations and Insights:**
- The above output shows that **the predicted ratings are higher than the corrected ratings** by the **user-item-Co-clustering based optimized**. This implies that the model is over-estimating the ratings.

### Content Based Recommendation Systems

**Think About It:** So far we have only used the play_count of songs to find recommendations but we have other information/features on songs as well. Can we take those song features into account?

In [66]:
df_small = df_final

In [67]:
#final_ratings['genres'] = final_ratings['genres'].apply(lambda x: " ".join(x.split('|')))

# Concatenate the "title", "release", "artist_name" columns to create a different column named "text"
df_small['text'] = df_small['title'] + ' ' + df_small['release'] + ' ' + df_small['artist_name']

In [68]:
# Select the columns 'user_id', 'song_id', 'play_count', 'title', 'text' from df_small data
df_small = df_small[['user_id', 'song_id', 'play_count', 'title', 'text']]

# Drop the duplicates from the title column
df_small = df_small.drop_duplicates(subset = ['title'])

# Set the title column as the index
df_small = df_small.set_index('title')

# See the first 5 records of the df_small dataset
df_small.head()

user_id  song_id  play_count  \
title                                                           
Daisy And Prudence                  6958      447           1   
The Ballad of Michael Valentine     6958      512           1   
I Stand Corrected (Album)           6958      549           1   
They Might Follow You               6958      703           1   
Monkey Man                          6958      719           1   

                                                                              text  
title                                                                               
Daisy And Prudence                    Daisy And Prudence Distillation Erin McKeown  
The Ballad of Michael Valentine  The Ballad of Michael Valentine Sawdust The Ki...  
I Stand Corrected (Album)        I Stand Corrected (Album) Vampire Weekend Vamp...  
They Might Follow You                They Might Follow You Tiny Vipers Tiny Vipers  
Monkey Man                           Monkey Man You Know I'm No Good Amy Winehouse

In [69]:
# Create the series of indices from the data
indices = pd.Series(df_small.index)

indices[ : 5]

0                 Daisy And Prudence
1    The Ballad of Michael Valentine
2          I Stand Corrected (Album)
3              They Might Follow You
4                         Monkey Man
Name: title, dtype: object

In [70]:
# Importing necessary packages to work with text data
import nltk

# Download punkt library
nltk.download("punkt")

# Download stopwords library
nltk.download("stopwords")

# Download wordnet 
nltk.download("wordnet")

# find out what this is and why it is needed
nltk.download('omw-1.4')

# Import regular expression
import re

# Import word_tokenizer
from nltk import word_tokenize

# Import WordNetLemmatizer
from nltk.stem import WordNetLemmatizer

# Import stopwords
from nltk.corpus import stopwords



# Import CountVectorizer and TfidfVectorizer
from sklearn.feature_extraction.text import CountVectorizer, TfidfVectorizer

[nltk_data] Downloading package punkt to
[nltk_data]     C:\Users\rober\AppData\Roaming\nltk_data...
[nltk_data]   Package punkt is already up-to-date!
[nltk_data] Downloading package stopwords to
[nltk_data]     C:\Users\rober\AppData\Roaming\nltk_data...
[nltk_data]   Package stopwords is already up-to-date!
[nltk_data] Downloading package wordnet to
[nltk_data]     C:\Users\rober\AppData\Roaming\nltk_data...
[nltk_data]   Package wordnet is already up-to-date!
[nltk_data] Downloading package omw-1.4 to
[nltk_data]     C:\Users\rober\AppData\Roaming\nltk_data...
[nltk_data]   Package omw-1.4 is already up-to-date!


We will create a **function to pre-process the text data:**

In [71]:
# Function to tokenize the text
def tokenize(text):
    
    text = re.sub(r"[^a-zA-Z]"," ", text.lower())
    
    tokens = word_tokenize(text)
    
    words = [word for word in tokens if word not in stopwords.words("english")]  # Use stopwords of english
    
    text_lems = [WordNetLemmatizer().lemmatize(lem).strip() for lem in words]

    return text_lems

In [72]:
# Create tfidf vectorizer 
tfidf = TfidfVectorizer(tokenizer = tokenize)

# Fit_transfrom the above vectorizer on the text column and then convert the output into an array
small_tfidf = tfidf.fit_transform(df_small['text'].values).toarray()

In [73]:
pd.DataFrame(small_tfidf)

0     1     2     3     4     5     6     7     8     9     ...  1427  \
0     0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0  ...   0.0   
1     0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0  ...   0.0   
2     0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0  ...   0.0   
3     0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0  ...   0.0   
4     0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0  ...   0.0   
..    ...   ...   ...   ...   ...   ...   ...   ...   ...   ...  ...   ...   
556   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0  ...   0.0   
557   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0  ...   0.0   
558   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0  ...   0.0   
559   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0  ...   0.0   
560   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0  ...   0.0   

     1428  1429  1430  1431  1432  1433  1434  1435  1436  
0     0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0  
1     0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0  
2     0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0  
3     0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0  
4     0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0  
..    ...   ...   ...   ...   ...   ...   ...   ...   ...  
556   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0  
557   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0  
558   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0  
559   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0  
560   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0  

[561 rows x 1437 columns]

In [74]:
# Compute the cosine similarity for the tfidf above output
similar_play = cosine_similarity(small_tfidf, small_tfidf)
similar_play

array([[1., 0., 0., ..., 0., 0., 0.],
       [0., 1., 0., ..., 0., 0., 0.],
       [0., 0., 1., ..., 0., 0., 0.],
       ...,
       [0., 0., 0., ..., 1., 0., 0.],
       [0., 0., 0., ..., 0., 1., 0.],
       [0., 0., 0., ..., 0., 0., 1.]])

 Finally, let's create a function to find most similar songs to recommend for a given song.

In [75]:
# Function that takes in song title as input and returns the top 10 recommended songs
def recommendations(title, similar_songs):
    
    recommended_songs = []
    
    # Getting the index of the song that matches the title
    idx = indices[indices == title].index[0]

    # Creating a Series with the similarity scores in descending order
    score_series = pd.Series(similar_songs[idx]).sort_values(ascending = False)

    # Getting the indexes of the 10 most similar songs
    top_10_indexes = list(score_series.iloc[1 : 11].index)
    print(top_10_indexes)
    
    # Populating the list with the titles of the best 10 matching songs
    for i in top_10_indexes:
        recommended_songs.append(list(df_small.index)[i])
        
    return recommended_songs

Recommending 10 songs similar to Learn to Fly

In [76]:
# Make the recommendation for the song with title 'Learn To Fly'
recommendations("Learn To Fly", similar_play)

[509, 234, 423, 345, 394, 370, 371, 372, 373, 375]


['Everlong',
 'The Pretender',
 'Nothing Better (Album)',
 'From Left To Right',
 'Lifespan Of A Fly',
 'Under The Gun',
 'I Need A Dollar',
 'Feel The Love',
 'All The Pretty Faces',
 'Bones']

**Observations and Insights:_________**

## **Conclusion and Recommendations:** 

- **Refined Insights -** What are the most meaningful insights from the data relevant to the problem?

- **Comparison of various techniques and their relative performance -** How do different techniques perform? Which one is performing relatively better? Is there scope to improve the performance further?

- **Proposal for the final solution design -** What model do you propose to be adopted? Why is this the best solution to adopt?

In [109]:
precision_recall_at_k(sim_user_user)
print("")
precision_recall_at_k(sim_user_user_optimized)

RMSE: 1.0878
Precision:  0.396
Recall:  0.692
F_1 score:  0.504

RMSE: 1.0521
Precision:  0.413
Recall:  0.721
F_1 score:  0.525


In [110]:
precision_recall_at_k(sim_item_item)
print("")
precision_recall_at_k(sim_item_item_optimized)

RMSE: 1.0394
Precision:  0.307
Recall:  0.562
F_1 score:  0.397

RMSE: 1.0328
Precision:  0.408
Recall:  0.665
F_1 score:  0.506


In [111]:
precision_recall_at_k(svd)
print("")
precision_recall_at_k(svd_optimized)

RMSE: 1.0252
Precision:  0.41
Recall:  0.633
F_1 score:  0.498

RMSE: 1.0141
Precision:  0.415
Recall:  0.635
F_1 score:  0.502


In [113]:
precision_recall_at_k(clust_baseline)
print("")
precision_recall_at_k(clust_tuned)

RMSE: 1.0487
Precision:  0.397
Recall:  0.582
F_1 score:  0.472

RMSE: 1.0654
Precision:  0.394
Recall:  0.566
F_1 score:  0.465


In [114]:
precision_recall_at_k(sim_user_user_optimized)
print("")
precision_recall_at_k(sim_item_item_optimized)
print("")
precision_recall_at_k(svd_optimized)
print("")
precision_recall_at_k(clust_baseline)

RMSE: 1.0521
Precision:  0.413
Recall:  0.721
F_1 score:  0.525

RMSE: 1.0328
Precision:  0.408
Recall:  0.665
F_1 score:  0.506

RMSE: 1.0141
Precision:  0.415
Recall:  0.635
F_1 score:  0.502

RMSE: 1.0487
Precision:  0.397
Recall:  0.582
F_1 score:  0.472


**Observations and Insights:**
- The optimized models performed better than the baseline analogs, excluding the tuned clustering model.
- We have calculated **RMSE** to check **how far the overall predicted ratings** are from the **actual ratings**.
- Intuition of Recall - We are getting a **recall of almost 0.692**, which means out of **all the relevant songs, 69.2% are recommended**. 
- Intuition of Precision - We are getting a **precision of almost 0.396**, optimized user-user similarity model which means **out of all the recommended songs 39.6% are relevant**.
- Here **F_1 score** of the **baseline model is almost 0.504**. It indicates that **mostly recommended songs were relevant and relevant songs were recommended** to the user. 